File: LeetCode/python3/670.py
class Solution:
    def maximumSwap(self, num):
        """
        :type num: int
        :rtype: int
        """

        def supp(m):
            ret = 0
            while len(m) > 0:
                ret = ret*10 + m[-1]
                m = m[:len(m)-1]
            return ret

        m = []
        temp = num
        while temp != 0:
            m.append(temp % 10)
            temp //= 10
        latest = [-1 for i in range(10)]
        for i in range(len(m)-1,-1,-1):
            latest[m[i]] = i
        for i in range(len(m)-1,-1,-1):
            for j in range(9, m[i], -1):
                if latest[j] != -1 and i > latest[j]:
                    m[i], m[latest[j]] = m[latest[j]], m[i]
                    return supp(m)
        return num

File: LeetCode/cpp/670.cpp
class Solution {
public:
    int maximumSwap(int num) {
        vector<int> vec, latest(10, -1);
        int temp = num;
        while (num > 0) {
            int &&temp = num % 10;
            vec.push_back(temp);
            num /= 10;
        }
        auto size = vec.size();
        for (int i = size - 1; i >= 0; --i)
            latest[vec[i]] = i;
        for (int i = size - 1; i >= 0; --i) {
            for (int j = 9; j > vec[i]; --j) {
                if (latest[j] != -1 && i > latest[j]) {
                    swap(vec[i], vec[latest[j]]);
                    return support(vec);
                }
            }
        }
        return temp;
    }

    int support(vector<int> &vec) {
        int ret = 0;
        while (!vec.empty()) {
            ret = ret * 10 + vec.back();
            vec.pop_back();
        }
        return ret;
    }
};
